hilti/toolchain/src/compiler/visitors/validator.cc

```diff
@@ -107,12 +107,22 @@ struct VisitorPost : public hilti::visitor::PreOrder<void, VisitorPost>, public
 
     ////// Declarations
 
+    // Perform validation of ID names suitable for all types of declarations.
+    void checkDeclarationID(const ID& id, const position_t& p) {
+        if ( id.str() == "self" )
+            error("cannot use 'self' as identifier", p);
+    }
```

Comment:
Let's instead implement this as

```cpp
    void operator()(const Declaration& d, position_t p) {
        // Perform validation of ID names suitable for all types of declarations.
        if ( d.id().str() == "self" )
            error("cannot use 'self' as identifier", p);
    }
```

and remove the explicit calls. I believe this would catch additional declarations.

We could add more tests for that (e.g., funcion or module decls named `self`), but below test would then also report redefinition errors.

Reply:
Doesn't work unfortunately because `declaration::Expression` is internally used for mapping `self` to its unit instance. The validator code would trigger for that, too. We could white-list that, but not sure that's really better then.

Reply:
Hmm, I didn't see that.

The current implementation misses some declarations, e.g, `Module`, `Type`, or `Function`, so I think using a generic visitor over `Declaration` with whitelisting for `declaration::Expression` is still better (if only since it allows us to put a comment why we ignore `Expression` instead of no coding being there).

This still has issues visiting a `declaration::Module` and I was also unable to visit it explicitly. I am unsure why that is (such nodes would e.g., not be root nodes).

With a generic visitor all of these test cases but the last one should be rejected:

```ruby
module foo;
function self(xyz: uint8) {}
##########################################
module foo;
type self = struct {
	self: uint8;
};
##########################################
module foo;
type self = unit {
	self: uint8;
};
##########################################
module self;
```

Reply:
Ok, convinced (but not sure why `Module` doesn't trigger)